wgengine/router/router_func.go

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package router
+
+import (
+	"tailscale.com/types/logger"
+)
+
+// setRoutesFunc applies the given router settings to the OS network stack.
+type setRoutesFunc func(cfg *Config) error
+
+// funcRouter passes the configs it receives to setRoutesFunc.
+// It is useful, for example, to pass configs into ipn-go-bridge on macOS/iOS.
+type funcRouter struct {
+	setRoutesFunc setRoutesFunc
+}
+
+// NewFuncRouter returns a Router which passes the configs it receives to setRoutesFunc.
+func NewFuncRouter(logf logger.Logf, setRoutesFunc setRoutesFunc) (Router, error) {
+	return funcRouter{
+		setRoutesFunc: setRoutesFunc,
+	}, nil
+}
+
+func (funcRouter) Up() error {
+	// Bringing up the routes is handled externally.
```

Comment:
well, to make this more generally a "FuncRouter", perhaps the up func should also be provided in the constructor (optionally nil), then we don't have to make this statement here and make assumptions about our one caller.

Then this is

    if r.up = nil {
         return nil
    }
    return r.up()

Reply:
Made it more general by exposing the whole suite (up/down/set) of functions.